src/lib/reasoners/bitv.ml

```
@@ -289,40 +315,557 @@ module Shostak(X : ALIEN) = struct
 
-    let string_to_bitv s =
+    let string_to_bitv ?(neg = false) s ctx =
+      let bitf = if neg then (fun b -> not b) else (fun b -> b) in
       let tmp = ref[] in
-      String.iter(fun car -> tmp := (not @@ Char.equal car '0', 1)::(!tmp)) s;
+      let nctx = ref[] in
+      String.iter (
+        fun car ->
+          tmp := (not @@ Char.equal  car '0',1)::(!tmp)
+      ) s;
       let rec f_aux l acc = match l with
         | [] -> assert false
-        | [(b,n)] -> { sz = n ; bv = I_Cte b }::acc
-        | (true as b1, n)::(true, m)::r | (false as b1, n)::(false, m)::r ->
-          f_aux ((b1,n+m)::r) acc
+        | [(b,n)] -> { sz = n ; bv = I_Cte (bitf b) }::acc
+        | (b1,n)::(b2,m)::r when Bool.equal b1 b2 -> f_aux ((b1,n+m)::r) acc
         | (b1,n)::(b2,m)::r ->
-          (f_aux ((b2,m)::r)) ({ sz = n ; bv = I_Cte b1 }::acc)
+          (f_aux ((b2,m)::r)) ({ sz = n ; bv = I_Cte (bitf b1) }::acc)
       in
       let res = f_aux (!tmp) [] in
-      bitv_to_icomp (List.hd res) (List.tl res)
-
-    let make t =
-      let rec make_rec t' ctx = match E.term_view t' with
-        | { E.f = Sy.Bitv s; _ } -> string_to_bitv s, ctx
-        | { E.f = Sy.Op Sy.Concat ;
-            xs = [t1;t2] ; ty = Ty.Tbitv n; _ } ->
-          let r1, ctx = make_rec t1 ctx in
-          let r2, ctx = make_rec t2 ctx in
-          { bv = I_Comp (r1, r2) ; sz = n }, ctx
-        | { E.f = Sy.Op Sy.Extract (i, j); xs = [t1] ; ty = Ty.Tbitv _; _ } ->
-          let r1, ctx = make_rec t1 ctx in
-          { sz = j - i + 1 ; bv = I_Ext (r1,i,j)}, ctx
-
-        | { E.ty = Ty.Tbitv n; _ } ->
+      let ctx = List.rev_append !nctx ctx in
+      bitv_to_icomp (List.hd res) (List.tl res), ctx
+
+    let bit_of_expr t =
+      match E.term_view t with
+      | { f = Sy.Int s; _ } ->
+        begin match Hstring.view s with
+          | "1" -> Some true
+          | "0" -> Some false
+          | _ ->
+            failwith (
```

Use `Util.failwith` instead of `failwith` + `asprintf`

---

Done